File: backend/app/com/skewwhiffy/auraltester/clefs/ClefNotes.scala
package com.skewwhiffy.auraltester.clefs

import com.skewwhiffy.auraltester.notes.{AbsoluteNote, Accidental, Interval, Note}
import com.skewwhiffy.auraltester.scales.NoteSequence

import scala.util.chaining.scalaUtilChainingOps

trait ClefNotes extends NoteSequence {
  implicit class AbsoluteNoteExtensions(val source: AbsoluteNote) {
    def downOne: AbsoluteNote = (source - Interval.minor(2)).ignoreAccidental

    def upOne: AbsoluteNote = (source + Interval.minor(2)).ignoreAccidental

    def skipOne: AbsoluteNote = (source + Interval.minor(3)).ignoreAccidental

    def ignoreAccidental: AbsoluteNote = Note(source.note.noteName, Accidental.natural)
        .pipe(it => AbsoluteNote(it, source.octave, source.lyric))

    def withNoteName: AbsoluteNote = source.withLyric(source.note.displayString)
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/abc/ClefNotes.kt
package com.skewwhiffy.auraltester.notation.model.abc

import com.skewwhiffy.auraltester.notation.model.interval.Interval
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import com.skewwhiffy.auraltester.notation.model.note.Accidental
import com.skewwhiffy.auraltester.notation.model.note.Note
import com.skewwhiffy.auraltester.notation.model.note.NoteSequence
import com.skewwhiffy.auraltester.notation.model.note.displayString

interface ClefNotes : NoteSequence

val AbsoluteNote.downOne: AbsoluteNote
    get() = (this - Interval.minor(2)).ignoreAccidental

val AbsoluteNote.upOne: AbsoluteNote
    get() = (this + Interval.minor(2)).ignoreAccidental

val AbsoluteNote.skipOne: AbsoluteNote
    get() = (this + Interval.minor(3)).ignoreAccidental

val AbsoluteNote.ignoreAccidental: AbsoluteNote
    get() = AbsoluteNote(Note(note.noteName, Accidental.natural), octave, lyric)

val AbsoluteNote.withNoteName: AbsoluteNote
    get() = withLyric(note.displayString)